### DISCUSSION TITLE
How to collect code coverage with distributed tasks?

### DISCUSSION THREAD

**Discussion by alfaproject** [nrwl/nx#17359] [General]
We have a monorepo and for each branch of work we run only the tasks that were affected by that specific branch changes in a distributed manner using NX cloud. Bliss (:

But how do we know what's the code coverage for the whole monorepo in that particular branch if only 1 or 2 projects were tested instead of all of them?

What are folks doing to collect this metric in their project for each CI run?

**Comment by robertIsaac** [comment]
IMO since you already have nx cloud, you don't really need to run it on affected only, since the unaffected ones will be cached
I'm curious though how do you report the coverage for all the tasks?

**Reply by alfaproject** [reply]
We don't report the coverage for the whole repo. That was my question (:

By now, I was kind of hoping that kind of thing would be handled by NX, perhaps by storing the test reports and parse them and maybe display in NX cloud somewhere but I guess no one asked for it.

Never thought about running all the tasks for all the projects all the time but isn't that against the NX philosophy?

**Reply by robertIsaac** [reply]
> isn't that against the NX philosophy?

it depend if you enable cloud caching or not
if not, of course it's not okay, since it will be costly
if you are enabling it, it will just fetch the report from the cloud without executing them, which is exactly what you need

I mean imagine you have 10 projects, you made changes to only 2 and 8 are unchanged in a feature branch
that mean the other 8 projects will have the same hash as your main branch, since they are unchanged
so running test for all projects will fetch 8 from the cloud, and run the test for the 2 projects that has different hash

**Reply by alfaproject** [reply]
I see. Should this be updated to run all the projects all the time? https://nx.dev/recipes/ci/monorepo-ci-gitlab#distributed-ci-with-nx-cloud

**Reply by robertIsaac** [reply]
yes replace every `nx affected` with `nx run-many`
if you are using old version of nx you will need to add `--all`, but this is the default behavior going forward
also you should remove `--base=$NX_BASE --head=$NX_HEAD` since they are related to only affected command

**Reply by robertIsaac** [reply]
you can do it only for unit testing if you care only about the test coverage
to replace 
```
yarn nx affected --base=$NX_BASE --head=$NX_HEAD -t test --parallel=3 --configuration=ci
```
with
```
yarn nx run-many -t test --parallel=3 --configuration=ci
```
but I believe you have `--coverage` also
